### PR TITLE
[ml] Use temporary data reference SAS uri for snapshot uploads

### PR DESCRIPTION
**Pull request by diondrapeck**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

The main objective of this PR is to begin the switch from local upload requiring `list_keys` access (not granted to all users) to generating a write SAS using workspace storage MSI for uploading snapshots. Feature overview provided [here](https://github.com/Azure/azureml_run_specification/pull/867).

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ml/azure-ai-ml/azure/ai/ml/constants/_common.py [hunk 1]
```diff
@@ -67,6 +67,8 @@
     "Only AmlCompute/KubernetesCompute cluster properties are supported, compute name {}, is {} type."
 )
 MAX_AUTOINCREMENT_ATTEMPTS = 3
+MAX_ASSET_STORE_API_CALL_ATTEMPTS = 3
+REGISTRY_DISCOVERY_BASE_URI = "https://eastus.api.azureml.ms"
```
Comment (banibrata) [inline sdk/ml/azure-ai-ml/azure/ai/ml/constants/_common.py:71]: Consider using this : https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/ml/azure-ai-ml/azure/ai/ml/_azure_environments.py#:~:text=EndpointURLS.REGISTRY_DISCOVERY_ENDPOINT%3A%20%22https%3A//eastus.api.azureml.ms/%22%2C

Comment (diondrapeck) [reply sdk/ml/azure-ai-ml/azure/ai/ml/constants/_common.py:71]: Thanks so much!

File: sdk/ml/azure-ai-ml/azure/ai/ml/_artifacts/_artifact_utilities.py [hunk 2]
```diff
@@ -429,18 +441,243 @@ def _check_and_upload_env_build_context(
...
+    workspace_location = workspace.location
+
+    # build asset id
+    asset_id = f"azureml://locations/{workspace_location}/workspaces/{workspace_id}/"\
+        f"{asset_type}/{asset_name}/versions/{asset_version}"
+
+    # build and send request
+    data = {
+        "assetId": asset_id,
+        "temporaryDataReferenceId": temporary_data_reference_id,
```
Comment (andscho-msft) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_artifacts/_artifact_utilities.py:493]: From this spec, looks like they want to rename everything to pending upload id. Will there be a huge effort to rename it after this Pr ships? These generated codes can just be re-generated?

File: sdk/ml/azure-ai-ml/azure/ai/ml/_artifacts/_artifact_utilities.py [hunk 3]
```diff
@@ -4,12 +4,16 @@
...
 # pylint: disable=protected-access
 
+import json
 import logging
 import os
 import uuid
 from datetime import datetime, timedelta
 from pathlib import Path
 from typing import Dict, Optional, Tuple, TypeVar, Union
+import requests
```
Comment (needuv) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_artifacts/_artifact_utilities.py:14]: I don't think we're allowed to take a dependency on `requests`

File: sdk/ml/azure-ai-ml/azure/ai/ml/constants/_common.py [hunk 4]
```diff
@@ -165,6 +166,7 @@
     "Spark job will only install the packages defined in the Conda configuration. It "
     "will not create a docker container using the image defined in the environment."
 )
+SERVICE_URL = "https://{}.api.azureml.ms"
```
[the remark this answers is not in the document]
Comment (diondrapeck) [reply sdk/ml/azure-ai-ml/azure/ai/ml/constants/_common.py:174]: I'll update everything to use the appropriate service url based on the cloud type. Thank you for catching this!

File: sdk/ml/azure-ai-ml/azure/ai/ml/_artifacts/_artifact_utilities.py (hunk 2 above)
Comment (diondrapeck) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_artifacts/_artifact_utilities.py:494]: What do you mean by "rename everything"? Change the name of the key in the data dictionary?

File: sdk/ml/azure-ai-ml/azure/ai/ml/_artifacts/_artifact_utilities.py (hunk 3 above)
Comment (diondrapeck) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_artifacts/_artifact_utilities.py:14]: Thanks for bringing this to my attention. I talked with Kevin and have refactored to use the HttpPipeline class.

File: sdk/ml/azure-ai-ml/azure/ai/ml/constants/_common.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (diondrapeck) [reply sdk/ml/azure-ai-ml/azure/ai/ml/constants/_common.py]: addressed

**Pull request closed without merging**